src/Teleportation.tsx

```diff
@@ -0,0 +1,166 @@
+import * as React from 'react'
+import { useThree } from '@react-three/fiber'
+import { Interactive, useXR } from '@react-three/xr'
+import { useCallback, useEffect, useState } from 'react'
+import { Quaternion, Vector3 } from 'three'
+
+export function useTeleportation() {
+  const [baseReferenceSpace, setBaseReferenceSpace] = useState<XRReferenceSpace | null>(null)
+  const { xr: xrManager } = useThree((state) => state.gl)
+  const { session } = useXR()
+
+  useEffect(() => {
+    const b = xrManager.getReferenceSpace()
+    if (b) {
+      setBaseReferenceSpace(b)
+    }
+    const onSessionStart = () => {
+      if (xrManager) {
+        setBaseReferenceSpace(xrManager.getReferenceSpace())
+      }
+    }
+    xrManager.addEventListener('sessionstart', onSessionStart)
+    return () => {
+      xrManager.removeEventListener('sessionstart', onSessionStart)
+    }
+  }, [xrManager])
+
+  const teleportTo = useCallback(
+    async (worldPosition: Vector3) => {
+      if (baseReferenceSpace && session) {
+        const pose = await new Promise<XRViewerPose | undefined>((resolve) => {
+          session.requestAnimationFrame((_, xrFrame) => {
+            const pose = xrFrame.getViewerPose(baseReferenceSpace)
+            resolve(pose)
+          })
+        })
+
+        const offsetX = pose?.transform.position.x || 0
+        const offsetZ = pose?.transform.position.z || 0
+
+        const offsetFromBase = {
+          x: -worldPosition.x + offsetX,
+          y: -worldPosition.y,
+          z: -worldPosition.z + offsetZ
+        }
+        const transform = new XRRigidTransform(offsetFromBase, new Quaternion())
+        const teleportSpaceOffset = baseReferenceSpace.getOffsetReferenceSpace(transform)
+        xrManager.setReferenceSpace(teleportSpaceOffset)
```

Comment:
Hmm, I may not be the best person to answer this question about WebXR, but I think the reason for modifying XR reference space instead of player objects in WebXR is to handle weird device specific issues around player camera positioning. It will come come into play when we do anchors since they get poses relative to the base reference frame. :)